embassy-rp/src/pio.rs

```diff
@@ -306,157 +289,251 @@ impl<PIO: PioInstance> PioPin<PIO> {
     }
 
-    fn tx_level(&self) -> u8 {
+    pub fn stalled(&self) -> bool {
         unsafe {
-            let flevel = Self::Pio::PIO.flevel().read().0;
-            (flevel >> (Self::SM * 8)) as u8 & 0x0f
+            let fdebug = PIO::PIO.fdebug();
+            let ret = fdebug.read().txstall() & (1 << SM) != 0;
+            fdebug.write(|w| w.set_txstall(1 << SM));
+            ret
         }
     }
 
-    fn rx_level(&self) -> u8 {
+    pub fn overflowed(&self) -> bool {
         unsafe {
-            let flevel = Self::Pio::PIO.flevel().read().0;
-            (flevel >> (Self::SM * 8 + 4)) as u8 & 0x0f
+            let fdebug = PIO::PIO.fdebug();
+            let ret = fdebug.read().txover() & (1 << SM) != 0;
+            fdebug.write(|w| w.set_txover(1 << SM));
+            ret
         }
     }
 
-    fn push_tx(&mut self, v: u32) {
+    pub fn push(&mut self, v: u32) {
         unsafe {
-            Self::Pio::PIO.txf(Self::SM).write_value(v);
+            PIO::PIO.txf(SM).write_value(v);
         }
     }
 
-    fn try_push_tx(&mut self, v: u32) -> bool {
-        if self.is_tx_full() {
+    pub fn try_push(&mut self, v: u32) -> bool {
+        if self.full() {
             return false;
         }
-        self.push_tx(v);
+        self.push(v);
         true
     }
 
-    fn pull_rx(&mut self) -> u32 {
-        unsafe { Self::Pio::PIO.rxf(Self::SM).read() }
+    pub fn wait_push<'a>(&'a mut self, value: u32) -> FifoOutFuture<'a, 'd, PIO, SM> {
+        FifoOutFuture::new(self, value)
     }
 
-    fn try_pull_rx(&mut self) -> Option<u32> {
-        if self.is_rx_empty() {
-            return None;
+    pub fn dma_push<'a, C: Channel, W: Word>(&'a mut self, ch: PeripheralRef<'a, C>, data: &'a [W]) -> Transfer<'a, C> {
+        unsafe {
+            let pio_no = PIO::PIO_NO;
+            let p = ch.regs();
+            p.read_addr().write_value(data.as_ptr() as u32);
+            p.write_addr().write_value(PIO::PIO.txf(SM).ptr() as u32);
+            p.trans_count().write_value(data.len() as u32);
+            compiler_fence(Ordering::SeqCst);
+            p.ctrl_trig().write(|w| {
+                // Set TX DREQ for this statemachine
+                w.set_treq_sel(TreqSel(pio_no * 8 + SM as u8));
+                w.set_data_size(W::size());
+                w.set_chain_to(ch.number());
+                w.set_incr_read(true);
+                w.set_incr_write(false);
+                w.set_en(true);
+            });
+            compiler_fence(Ordering::SeqCst);
+        }
+        Transfer::new(ch)
+    }
+}
+
+pub struct PioStateMachine<'d, PIO: PioInstance, const SM: usize> {
+    rx: PioStateMachineRx<'d, PIO, SM>,
+    tx: PioStateMachineTx<'d, PIO, SM>,
+}
+
+impl<'d, PIO: PioInstance, const SM: usize> Drop for PioStateMachine<'d, PIO, SM> {
+    fn drop(&mut self) {
+        unsafe {
+            PIO::PIO.ctrl().write_clear(|w| w.set_sm_enable(1 << SM));
+        }
+        on_pio_drop::<PIO>();
```

Comment:
dropping one state machine will deconfigure all PIO pins. I think this could be a footgun, if you only use one SM, and drop all the others early?